design/API_requirements_and_design.md

```diff
@@ -0,0 +1,59 @@
+# Thin-Edge API requirements and design
+
+One should be able to build a thin-edge executable from tedge plugins crates that have been implemented independently:
+
+* All these tedge plugins crates depends on the tedge_api crate that defines how
+  to create, connect and run plugin instances.
+* Each tedge plugin crate can define its own types of messages. For instance, a
+  tedge_telemetry plugin can defines measurements, events, alarms and setpoints
+  messages.
+* To exchange typed messages, plugin crates need dependency relationship. For
+  instance, the tedge_json_over_mqtt plugin crate will have to depend on the
+  tedge_telemetry plugin crate and the tedge_mqtt crate.
```

Comment:
I am not sure we should even write this in the document. Because to exchange typed messages, plugins only need to know the type, but not whom it may come from.

Think of this:

* A crate `telemetry_types` defines `Measurement`
* A plugin `plugin_c8y` can send `Measurement`s
* A plugin `plugin_measurement_to_json` can receive `Measurement`s, turn it into `Vec<u8>` and send that
* A plugin `plugin_mqtt` can receive `Vec<u8>` and send that to an MQTT broker

None of the above plugins actually need to know that the other plugins actually exist! Only `plugin_c8y` and `plugin_measurement_to_json` need to depend on `telemetry_types`, but not even on eachother!

So maybe we should rephrase that paragraph, or leave it out if we cannot think of something else?

Reply:
We are fully aligned!

The sentence might be rewritten using the details you gave in your comment.

Reply:
Added a fixup commit to address this, but feel free to suggest more, I kept the point rather short.